### DISCUSSION TITLE
need help mapping external field to nodes

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#21371] [Q&A General] [answered]
Hi all -

I'm trying to apply an external field (temperature) to nodes on my mesh. I've got a list of node numbers and/or nodesets and the required temperature, which can also vary through time. I want to read in the values and apply them at the nodes, and also be able to interpolate across the element to the quadrature points for my material model. I figured a Nodal AuxKernel is the way to go for the last part, but what's a good way to read in nodal values, that change with time (piecewise linearly), from a file?

**Comment by GiudGiud** [comment, marked as answer]
Hello

The only time-dependent CSV reader I know of is using a Function.
https://mooseframework.inl.gov/source/functions/PiecewiseMulticonstant.html
and
https://mooseframework.inl.gov/source/functions/PiecewiseMultilinear.html

this does not store info by nodes though, only XYZT

so what you want here would be a time-dependent PropertyReadeFile user object. 
We dont have that but we could use one
https://mooseframework.inl.gov/source/userobject/PropertyReadFile.html

Guillaume

**Reply by jessecarterMOOSE** [reply]
Thanks. Let's say I manage to get all the data read in. How do I get those values to the nodes and then to the quadrature points? Is there an AuxKernel that calls the property data?

**Reply by GiudGiud** [reply]
so the way we do it with the propertyReadFile is that there is a different read routine/workflow depending on whether the data is stored by node / element / block / nearest-neighbor region. 

I would imagine you do the same, but add a time / time_index argument to the read routine. 

Yes you may use an AuxKernel. We have not made a XXXXXFromCSVAuxKernel but it wouldnt be hard to make, following the same patterns as the PiecewiseConstantFromCSV Function
For getting data to nodes just store it by nodes and read by nodes.
For Qps, either do the same (store by Qp read by Qp) or the usual workflow does the conversion node-qp qp-nodes with the shape functions and what not.
